pkg/reporting/preflight_airgap.go

```diff
@@ -1,29 +1,117 @@
 package reporting
 
 import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
 	"github.com/pkg/errors"
-	reportingtypes "github.com/replicatedhq/kots/pkg/api/reporting/types"
 	"github.com/replicatedhq/kots/pkg/buildversion"
-	"github.com/replicatedhq/kots/pkg/store"
+	"github.com/replicatedhq/kots/pkg/k8sutil"
 	storetypes "github.com/replicatedhq/kots/pkg/store/types"
+	"github.com/replicatedhq/kots/pkg/util"
 	kotsv1beta1 "github.com/replicatedhq/kotskinds/apis/kots/v1beta1"
+	kuberneteserrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	PreflightReportSecretNameFormat = "kotsadm-%s-preflight-report"
+	PreflightReportSecretKey        = "report"
+	PreflightReportEventLimit       = 4000
 )
 
+var preflightReportMtx = sync.Mutex{}
+
 func (r *AirgapReporter) SubmitPreflightData(license *kotsv1beta1.License, appID string, clusterID string, sequence int64, skipPreflights bool, installStatus storetypes.DownstreamVersionStatus, isCLI bool, preflightStatus string, appStatus string) error {
-	status := &reportingtypes.PreflightStatus{
+	app, err := r.store.GetApp(appID)
+	if err != nil {
+		if r.store.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to get airgapped app")
+	}
+
+	event := PreflightReportEvent{
+		ReportedAt:      time.Now().UTC().UnixMilli(),
+		LicenseID:       license.Spec.LicenseID,
 		InstanceID:      appID,
 		ClusterID:       clusterID,
 		Sequence:        sequence,
 		SkipPreflights:  skipPreflights,
 		InstallStatus:   string(installStatus),
 		IsCLI:           isCLI,
 		PreflightStatus: preflightStatus,
-		AppStatus:       preflightStatus,
-		KOTSVersion:     buildversion.Version(),
+		AppStatus:       appStatus,
+		KotsVersion:     buildversion.Version(),
+	}
+
+	if err := CreatePreflightReportEvent(r.clientset, util.PodNamespace, app.Slug, event); err != nil {
+		return errors.Wrap(err, "failed to create preflight report event")
+	}
+
+	return nil
+}
+
+func CreatePreflightReportEvent(clientset kubernetes.Interface, namespace string, appSlug string, event PreflightReportEvent) error {
```

Comment:
Thoughts on refactoring this function and `CreateInstanceReportEvent` so that both the instance and preflight reports use the same function? I did a smaller DRY for some of the utility functions, but thought that having this function accept a `report any` might make it somewhat less understandable.

Reply:
Yeah. You can create a `Report` interface, and maybe move all of this logic to `report.go` in this `reporting` package. You can then implement functions like `GetEventLimit`, `GetSecretKey`, etc...

You can then have a single `CreateReportEvent` function that accepts this interface. I'd also suggest renaming `AirgapReportSecret` to `ReportSecret`, `DecodeAirgapReport` to `DecodeAReport`, `EncodeAirgapReport` to `EncodeReport`.